Source/DafnyStandardLibraries/src/dafnystdlibs/Wrappers.md

```diff
@@ -0,0 +1,191 @@
+
+## The `Wrappers` module
+
+The Wrappers module holds some
+simple datatypes to support common patterns, such as optional values or the result of operations that can fail.
+These are additionally useful with Dafny's abrupt-termination-on-failure `:-` operator.
+
+Any user datatype can serve this purpose, as long as it has an `IsFailure?` predicate 
+(in which case it is known as a [_failure-compatible_](https://dafny.org/latest/DafnyRef/DafnyRef#sec-failure-compatible-types) type --- an FC-type). 
+
+In this library module Dafny defines three such types:
+- `Option<R>` - which is either `Some` with a value of type `T` or a `None` with no information
+- `Outcome<E>` - which is either `Pass` with no informatino or `Fail` with an error value of type `E` (often a `string`) 
+- `Result<R,E>` - which is either `Success` with a value of type `R` or `Failure` with error value of type `E`
+
+These are common programming idioms. The main complication comes when they are mixed in the same program.
+
+### Option
+
+Consider this routine that looks for a value in a sequence, beginning at position `k`, returning its index:
+<!-- %check-verify %save tmp-find.dfy -->
+```dafny
+import opened DafnyStdLibs.Wrappers
+function Find<T(==)>(s: seq<T>, k: int, value: T): (r: Option<int>)
+  requires 0 <= k <= |s|
+  ensures  r.Some? ==> 0 <= r.Extract() < |s| && s[r.Extract()] == value
+  ensures  r.None? ==> forall i | k <= i < |s| :: s[i] != value
+  decreases |s| - k
+{
+  if k >= |s| then None else if s[k] == value then Some(k) else Find(s, k+1, value)
+}
+```
+
+It could be used in a method like this
+
+<!-- %check-resolve %use tmp-find.dfy -->
+```dafny
+method m(s: seq<int>) returns (r: Option<int>) {
+  var value: int;
+  // do various calculations
+  var index: int :- Find(s, 0, value);
+  // more calculations
+  return Some(value);
+}
+```
+
+You could just capture the result of `Find` using `:=` in a `Option<int>` variable and inspect it. But if the `None` condition is 
+generally a rare error, it is easy to forget to always check that each operation was successful. Instead, the `:-` changes the 
+control flow so that if a `None` value is returned from `Find`, the method immediately aborts, with the output value (which has
+the `Option<int>` type) getting that returned `None` value. So this operates something like exceptions.
+See the [reference manual](https://dafny.org/latest/DafnyRef/DafnyRef#sec-update-with-failure-statement)
+for more on the similarities and differences with exceptions.
+
+### Outcome
```

Comment:
Is this used a lot? C#'s anonymous datatypes `Option<(int x, string y)>` seem nicer

Reply:
I actually disagree with the statement that this is the most common use case. It's more important that returning an Outcome instead of a Result means you can say `:- CheckSomeThings(x)` as opposed to `var _ :- CheckSomeThings(x)`, where in the latter case it's too easy to do `var _ := CheckSomeThings(x)` which has no effect!

Rewriting.

Reply:
Ended up just flipping the order of use cases.